### ISSUE TITLE
Unexpected behaviour while changing sources array

### ISSUE THREAD

**Issue by IlyaSmagin** [thejsn/react-image-mosaic#3]
I wanted to create a feature for my [collage website](playlist-collage.vercel.app), ([repo](https://github.com/IlyaSmagin/playlist-collage)) for a usecase where:

### I want to rerender ReactImageMosaic component with the same props but different sources arrays, provided by user.

While implementing i found out that I can't do that, because while **changing sources arrays I faced an unexpected ReactImageMosaic component behavior:**
- Images from removed array stayed in a collage
- Only some images from new source array appeared in a collage

But for me expected behavior is:
- All images from removed array should be removed from the collage
- New collage should be generated with images only from new source array

### I reproduced my issue on [CodeSandbox](https://codesandbox.io/s/reactimagemosaic-sources-bug-reproduction-ogsywk?file=/src/App.tsx)

Also I looked on the ReactImageMosaic codebase and found out possible TODO comment with probably similar functionality:
[src/utils/sources.ts](https://github.com/thejsn/react-image-mosaic/blob/master/src/utils/sources.ts) at line 14

### Is behavior I'm getting expected or is it a bug?

If this is a bug I would like to contribute, but I need to be more familiar with codebase...

And if this is expected and everything is working correctly, what can I do to get my desired result of complete replacement of source array?

**Comment by thejsn** [comment]
Hello!
Thanks for putting in the effort and write a detailed issue, I had forgotten about this! This is sort of a known bug, I should probably have done a better job documenting it. Also great that you want to contribute! I'd be happy to have a look at a PR and give feedback.

Here is some info about the issue that might help:

### The underlying issue

When the source array updates, the new images are added to `Grid.ts`. Existing images are never removed though. This was to avoid downloading images that has already been loaded when the source array updates.

The `updateSource` function in `utils/sources.ts` removes duplicates before calling `addSourcesFromURL`, but doesn't do anything with the images that has been removed. The TODO you found there was a start at a solution to this issue but i didn't have time to finish it.

In `Grid.ts` there are `_colors` and `_pictures` that need to be updated when sources updates. Each image that was removed from sources must be removed from `_pictures`, as well as the corresponding color in `_colors`, to solve the issue. 

### Possible solutions

One solution is to create a `removeSourceImage(url: string)` method in `Grid.ts` and call it for each url in `toRemove` from the unfinished TODO. In this new method go through the `_pictures` Map and find the item where the image source is equal to the passed url and remove it. To get the image source, add a new getter function in the Picture class that returns the source of `_image`. The matching color (the key in the map) must also be removed from _colors. 

Another possible solution if the first doesn't work is to simply clear all images and colors and reload all images when the sources updates. The browser is probably(?) caching the images anyway so this might be an unnecessary optimization(?). To do this, just create a `reset()` method in `Grid.ts` and clear `_colors` and `_pictures` in it, and call it before adding the new ones.

**Comment by IlyaSmagin** [comment]
#4 I think this PR fixes the issue

Tested on local demo, and now changing arrays works fine

Your guidance was greatly helpful, and now I'm more familiar with the repo, so thanks a lot!

Sorry for formatting issues, text editors are too intrusive nowadays...

Also this solution has nested loops (for each URL to remove we iterate through _pictures Map again), so it might be possible to optimize it if it will be a performance bottleneck. 
But I do not think this is the case as `removeSourcesByURL` works in parallel with `addSourcesFromURL` as `addSourcesFromURL` is async (at least I think so, as it returns Promise?), but if not, nevermind

I also could add changes to demo files to showcase that source array swap is now possible, or add a dark theme to the demo, as sometimes I work at night and it saved my eyes a lot

Please let me know if more changes to the PR needed!

**Comment by thejsn** [comment]
Hi ok so this is merged and published now in v2.2.0.
You can do another PR with the changes to the demo if you like.
